File: src/public/js/domains/util/scriptdat/format/codec.ts
import assert from 'assert';

const KEY = 0b01100001;

const CODE_MAP: ReadonlyArray<Readonly<{
  code: number;
  char: string;
}>> = (
  // tslint:disable-next-line:prefer-template
  '０１２３４５６７８９\nＢＣＤＥＦ'
  + 'ＳｄＯ新⑩倍母天道書者闇死地古文'
  + ` !"#$%&'()*+,-./`
  + '0123456789:;<=>?'
  + '@ABCDEFGHIJKLMNO'
  + 'PQRSTUVWXYZ[\\]^_'
  + '`abcdefghijklmno'
  + 'pqrstuvwxyz{|}~代'
  + '形勇気年杯体をぁぃぅぇぉゃゅょっ'
  + '真あいうえおかきくけこさしすせそ'
  + '実｡｢｣､･ｦｧｨｩｪｫｬｭｮｯ'
  + 'ｰｱｲｳｴｵｶｷｸｹｺｻｼｽｾｿ'
  + 'ﾀﾁﾂﾃﾄﾅﾆﾇﾈﾉﾊﾋﾌﾍﾎﾏ'
  + 'ﾐﾑﾒﾓﾔﾕﾖﾗﾘﾙﾚﾛﾜﾝﾞﾟ'
  + 'たちつてとなにぬねのはひふへほま'
  + 'みむめもやゆよらりるれろわん我▼'
)
  .split('')
  .map((x, i) => ({
    code: i,
    char: x,
  }));

const CHAR_TO_CODE: { [char: string]: number }
  // tslint:disable-next-line:prefer-object-spread
  = CODE_MAP.reduce((p, c) => ({ ...p, [c.char]: c.code }), {});
const CODE_TO_CHAR: { [code: number]: string }
  // tslint:disable-next-line:prefer-object-spread
  = CODE_MAP.reduce((p, c) => Object.assign(p, { [c.code]: c.char }), {});

export function decode(bin: ArrayBuffer) {
  let str = '';
  for (const item of new Uint8Array(bin)) {
    str += CODE_TO_CHAR[item ^ KEY];
  }
  return str;
}

export function encode(txt: string, buf: Uint8Array) {
  for (let i = 0; i < txt.length; i += 1) {
    buf[i] = CHAR_TO_CODE[txt.charAt(i)] ^ KEY;
  }
}

export function textToShopData(text: string) {
  return Uint8Array.from(
    text
      .split('')
      .map(x => CHAR_TO_CODE[x]),
  );
}

export function shopItemDataToText(shopItemData: Uint8Array) {
  assert.equal(shopItemData.length, 7 * 3);
  return [...shopItemData].map(x => CODE_TO_CHAR[x]).join('');
}

File: src-tauri/src/util/scriptdat/format/codec.rs
use std::collections::HashMap;

const KEY: u8 = 0b01100001;
const CODE_MAP: &str = concat!(
    "０１２３４５６７８９\nＢＣＤＥＦ",
    "ＳｄＯ新⑩倍母天道書者闇死地古文",
    " !\"#$%&'()*+,-./",
    "0123456789:;<=>?",
    "@ABCDEFGHIJKLMNO",
    "PQRSTUVWXYZ[\\]^_",
    "`abcdefghijklmno",
    "pqrstuvwxyz{|}~代",
    "形勇気年杯体をぁぃぅぇぉゃゅょっ",
    "真あいうえおかきくけこさしすせそ",
    "実｡｢｣､･ｦｧｨｩｪｫｬｭｮｯ",
    "ｰｱｲｳｴｵｶｷｸｹｺｻｼｽｾｿ",
    "ﾀﾁﾂﾃﾄﾅﾆﾇﾈﾉﾊﾋﾌﾍﾎﾏ",
    "ﾐﾑﾒﾓﾔﾕﾖﾗﾘﾙﾚﾛﾜﾝﾞﾟ",
    "たちつてとなにぬねのはひふへほま",
    "みむめもやゆよらりるれろわん我▼"
);

fn create_code_to_char() -> HashMap<u8, char> {
    return CODE_MAP
        .chars()
        .enumerate()
        .map(|(i, x)| (i as u8, x))
        .collect::<HashMap<_, _>>();
}

fn create_char_to_code() -> HashMap<char, u8> {
    return CODE_MAP
        .chars()
        .enumerate()
        .map(|(i, x)| (x, i as u8))
        .collect::<HashMap<_, _>>();
}

pub fn decode(from: &[u8]) -> String {
    let code_to_char = create_code_to_char();
    debug_assert_ne!(from[0], 0);
    let to: String = (0..from.len())
        .map(|i| {
            let query = from[i] ^ KEY;
            code_to_char[&query]
        })
        .collect();
    debug_assert_eq!(to.chars().next(), to.chars().nth(16));
    debug_assert_ne!(to.chars().next(), to.chars().nth(1));
    to
}

pub fn encode(from: &str) -> Vec<u8> {
    let char_to_code = create_char_to_code();
    from.chars()
        .map(|c| char_to_code[&c])
        .map(|ascii| ascii ^ KEY)
        .collect()
}

pub fn text_to_shop_data(text: &str) -> Vec<u8> {
    let char_to_code = create_char_to_code();
    text.chars().map(|c| char_to_code[&c]).collect()
}

pub fn shop_item_data_to_text(shop_item_data: &[u8]) -> String {
    debug_assert_eq!(shop_item_data.len(), 7 * 3);
    shop_item_data
        .iter()
        .map(|&x| CODE_MAP.chars().nth(x as usize).unwrap())
        .collect()
}
